src/btree/bt_cursor.c

```diff
@@ -2539,74 +2604,60 @@ __cursor_range_selectivity(WT_CURSOR_BTREE *start, WT_CURSOR_BTREE *stop, double
-        percentile_stop += ((double)(indx_stop - 1) / (double)(pindex_stop->entries - 1)) * selectivity_stop_node;
-
-        selectivity_start_node *= (double)1 / (double)(pindex_start->entries - 1);
-        selectivity_stop_node *= (double)1 / (double)(pindex_stop->entries - 1);
+          ((double)(indx_start - 1) / (double)start_child_count) * selectivity_start_node;
+        percentile_stop +=
+          ((double)(indx_stop - 1) / (double)stop_child_count) * selectivity_stop_node;
 
-        if (indx_start != indx_stop) {
-            diverged = true;
-        }
-
-        /*
-         * Swap the current page(s) for the child page(s). If the page splits while we're retrieving
-         * it, restart the search at the root.
-         */
-        read_flags = WT_READ_RESTART_OK;
-        if (F_ISSET(start, WT_CBT_READ_ONCE))
-            FLD_SET(read_flags, WT_READ_WONT_NEED);
-        if ((ret = __wt_page_swap(session, current_start, descent_start, read_flags)) == 0) {
-            current_start = descent_start;
-            if ((ret = __wt_page_swap(session, current_stop, descent_stop, read_flags)) == 0) {
-                current_stop = descent_stop;
-                continue;
-            }
-        }
-        if (ret == WT_RESTART)
-            goto restart;
-        return (ret);
+        selectivity_start_node *= (double)1 / (double)start_child_count;
+        selectivity_stop_node *= (double)1 / (double)stop_child_count;
     }
 
     // We reached a leaf node.
-    start_leaf_count = current_start->page->entries;
-    stop_leaf_count = current_stop->page->entries;
-    if (start_leaf_count == 0 || stop_leaf_count == 0) {
-        WT_ERR_MSG(session, WT_ERROR, "Found a leaf without entries. Did you forget to checkpoint?");
+    start_child_count = current_start->page->entries;
+    if (diverged) {
+        stop_child_count = current_stop->page->entries;
+    } else {
+        stop_child_count = start_child_count;
     }
 
-    ret = __cursor_range_selectivity_get_leaf_idx(&kstart, session, collator, start, page_start, current_start, &indx_start);
+    if (start_child_count == 0 || stop_child_count == 0) {
+        WT_ERR_MSG(
+          session, WT_ERROR, "Found a leaf without entries. Did you forget to checkpoint?");
+    }
+
+    ret = __cursor_range_selectivity_get_leaf_idx(
+      &kstart, session, collator, start, current_start, &indx_start);
     if (ret == WT_RESTART) {
         goto restart;
     }
     WT_ERR(ret);
 
-    if (!diverged) {
-        ret = __cursor_range_selectivity_get_leaf_idx(&kstop, session, collator, start, page_start, current_start, &indx_stop);
+    // At this point, we need to clear out the current pointers which have been moved
+    // into the btree cursor, so that cleanup never releases twice.
+    if (diverged) {
+        ret = __cursor_range_selectivity_get_leaf_idx(
+          &kstop, session, collator, stop, current_stop, &indx_stop);
+        current_stop = NULL;
```

Comment:
1. How'd you know to set it equal to NULL?
2. Shouldn't we also clear out the start pointer here?

Reply:
This is what I was messaging about on slack; it's [this line](https://github.com/HanaPearlman/wiredtiger/blob/0c9fc7745c790c96d1e8343a253852005066ac5e/src/btree/bt_cursor.c#L2251-L2255) in the helper that's supposed to be responsible for clearing this out. But the pointer is passed by reference, so it needs to also happen outside of the function!

> Shouldn't we also clear out the start pointer here?

It doesn't appear to be required and I'm not sure why! We definitely clean it up in the end (see `__wt_page_release`), but for some reason we don't get the double-free issue...